File: Solution/0081.Search_in_Rotated_Sorted_Array_II/0081.Search_in_Rotated_Sorted_Array_II.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int i=0,j=nums.size();
        int m;
        while(i<j)
        {
            m=i+(j-i)/2;
            if(nums[m]==target) return true;
            if(nums[i]>nums[m])
            {
                //右边非降序
                if(nums[m]<target&&nums[j-1]>=target) i=m+1;
                else j=m;
            }
            else if(nums[i]<nums[m])
            {
                if(nums[m]>target&&nums[i]<=target) j=m;
                else i=m+1;
            }
            else
            {
                while(i<j&&nums[i]==nums[m])
                    ++i;
                while(i<j&&nums[m]==nums[j-1])
                    --j;
            }

        }
        return false;
        
    }
};

File: Solution/0081.Search_in_Rotated_Sorted_Array_II/0081.Search_in_Rotated_Sorted_Array_II.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        i,j=0,len(nums)
        while i<j:
            m=i+(j-i)//2
            if nums[m]==target: return True
            if nums[m]>nums[i]:
                if target>=nums[i] and target<nums[m]:
                    j=m
                else:
                    i=m+1
            elif nums[m]<nums[i]:
                if target>nums[m] and target <=nums[j-1]:
                    i=m+1
                else:
                    j=m
            elif nums[m]==nums[i]:
                while i<j and nums[i] == nums[m]:
                    i+=1
                while i<j and nums[j-1] ==nums[m]:
                    j-=1
        return False
